### PR TITLE
feat: add solflare snap connect and signer

### PR DESCRIPTION
**Pull request by RyukTheCoder**
# Summary

Added Solflare Snap connect and signer.

Fixes # 1383

# How did you test this change?

Tested by connecting to Solflare Snap and signing a transaction.

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: wallets/provider-solflare-snap/package.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    "build": "node ../../scripts/build/command.mjs --path wallets/provider-metamask",
+    "ts-check": "tsc --declaration --emitDeclarationOnly -p ./tsconfig.json",
+    "clean": "rimraf dist",
+    "format": "prettier --write '{.,src}/**/*.{ts,tsx}'",
+    "lint": "eslint \"**/*.{ts,tsx}\" --ignore-path ../../.eslintignore"
+  },
+  "dependencies": {
+    "@rango-dev/signer-solana": "^0.26.0",
+    "@rango-dev/wallets-shared": "^0.30.0",
+    "@solana/web3.js": "^1.67.2",
```
Comment (RanGojo) [inline wallets/provider-solflare-snap/package.json:26]: I think you could remove dependency to solana/web3 and use our signer for handling transactions

File: wallets/provider-all/src/index.ts [hunk 2]
```diff
@@ -59,5 +60,6 @@ export const allProviders = (enviroments?: Enviroments) => {
     frontier,
     taho,
     braavos,
+    solflareSnap,
```
Comment (RanGojo) [inline wallets/provider-all/src/index.ts:63]: Please change the code, so I could see solana snap in connect wallet right after the metamask icon.

**Comment by RanGojo** [comment]
Good job @RyukTheCoder, let's finalize this PR for this spirit.

File: wallets/provider-solflare-snap/src/signers/solanaSigner.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+          versionedTransaction
+        );
+        console.log('result', result);
+      }
+      const hash = '';
+      return { hash };
+    } catch (error) {
+      throw new SignerError(SignerErrorCode.SEND_TX_ERROR, undefined, error);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (RanGojo) [reply wallets/provider-solflare-snap/src/signers/solanaSigner.ts:42]: Please sync with next. We had some updates for this function.

File: wallets/provider-solflare-snap/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+  };
+
+  instance.on('disconnect', async () => disconnect());
+
+  instance.on(
+    'accountChanged',
+    (publicKey: string) => async () => handleAccountsChanged(publicKey)
+  );
+};
```
[the remark this answers is not in the document]
Comment (RyukTheCoder) [reply wallets/provider-solflare-snap/src/index.ts]: I could not trigger "accountChanged" so I removed it.

File: wallets/provider-solflare-snap/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+export const getSigners: (provider: any) => SignerFactory = signer;
+
+export const canEagerConnect: CanEagerConnect = async ({ instance }) => {
+  return instance.standardAccounts?.length > 0;
+};
```
[the remark this answers is not in the document]
Comment (RyukTheCoder) [reply wallets/provider-solflare-snap/src/index.ts]: It seems that connection will be closed after refreshing the page. So I removed it.

File: wallets/provider-solflare-snap/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      CHROME:
+        'https://chrome.google.com/webstore/detail/metamask/nkbihfbeogaeaoehlefnkodbefgpgknn?hl=en',
+      BRAVE:
+        'https://chrome.google.com/webstore/detail/metamask/nkbihfbeogaeaoehlefnkodbefgpgknn?hl=en',
+
+      FIREFOX: 'https://addons.mozilla.org/en-US/firefox/addon/ether-metamask',
+      EDGE: 'https://microsoftedge.microsoft.com/addons/detail/metamask/ejbalbakoplchlghecdalmeeeajnimhm?hl=en-US',
+      DEFAULT: 'https://metamask.io/download/',
+    },
+    color: '#dac7ae',
```
[the remark this answers is not in the document]
Comment (RyukTheCoder) [reply wallets/provider-solflare-snap/src/index.ts]: Fixed

File: wallets/provider-solflare-snap/src/signers/solanaSigner.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (RyukTheCoder) [reply wallets/provider-solflare-snap/src/signers/solanaSigner.ts]: Due to some malicious warnings, decided to keep it as it is.

File: wallets/provider-solflare-snap/package.json (hunk 1 above)
Comment (RyukTheCoder) [reply wallets/provider-solflare-snap/package.json]: Done

File: wallets/provider-solflare-snap/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+export const getSigners: (provider: any) => SignerFactory = signer;
+
+export const getWalletInfo: (allBlockChains: BlockchainMeta[]) => WalletInfo = (
+  allBlockChains
+) => {
+  const solana = solanaBlockchain(allBlockChains);
+  return {
+    name: 'Solflare Snap',
+    img: 'https://raw.githubusercontent.com/rango-exchange/assets/main/wallets/metamask/icon.svg',
```
Comment (RanGojo) [inline wallets/provider-solflare-snap/src/index.ts:64]:
```suggestion
    img: 'https://raw.githubusercontent.com/rango-exchange/assets/main/wallets/solflare-snap/icon.svg',
```

File: wallets/provider-solflare-snap/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+export const getSigners: (provider: any) => SignerFactory = signer;
+
+export const getWalletInfo: (allBlockChains: BlockchainMeta[]) => WalletInfo = (
+  allBlockChains
+) => {
+  const solana = solanaBlockchain(allBlockChains);
+  return {
+    name: 'Solflare Snap',
```
Comment (RanGojo) [inline wallets/provider-solflare-snap/src/index.ts:63]:
```suggestion
    name: 'Solana Snap',
```

**Review by RanGojo — COMMENTED** [review]
LGTM, I've added some comments for finalizing this PR.

File: wallets/provider-solflare-snap/src/signers/solanaSigner.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+  async signAndSendTx(tx: SolanaTransaction): Promise<{ hash: string }> {
+    try {
+      const connection = getSolanaConnection();
+      const latestBlock = await connection.getLatestBlockhash('confirmed');
+
+      const finalTx = prepareTransaction(tx, latestBlock.blockhash);
+
+      const hash = await this.provider.signAndSendTransaction(finalTx);
```
Comment (RanGojo) [inline wallets/provider-solflare-snap/src/signers/solanaSigner.ts:30]: Please refactor and export this logic in solana-signer and call it here:

```ts
async function simulate (tx) {
  // We first simulate whether the transaction would be successful
  if (tx.txType === 'VERSIONED') {
    const { value: simulatedTransactionResponse } =
      await connection.simulateTransaction(
        finalTx as VersionedTransaction,
        {
          replaceRecentBlockhash: true,
          commitment: 'processed',
        }
      );
    const { err, logs } = simulatedTransactionResponse;
    if (err) {
      /*
       * Simulation error, we can check the logs for more details
       * If you are getting an invalid account error, make sure that you have the input mint account to actually swap from.
       */
      console.error('Simulation Error:', { err, logs });
      const message =
        (logs?.length || 0) > 0
          ? logs?.[logs?.length - 1]
          : JSON.stringify(err);
  
      throw new SignerError(
        SignerErrorCode.SEND_TX_ERROR,
        undefined,
        `Simulation failed: ${message}`
      );
    }
  }
}
```

and call it here:

```suggestion
      await simulateTransaction(finalTx);
      const hash = await this.provider.signAndSendTransaction(finalTx);
```

also you need to update your catch clause like this:

```ts
    } catch (error) {
      if (error instanceof SignerError) throw error;
      throw new SignerError(SignerErrorCode.SEND_TX_ERROR, undefined, error);
    }
```

**Review by RanGojo — APPROVED** [review]

File: wallets/provider-solflare-snap/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+};
+
+export const disconnect: Disconnect = async ({ instance }) => {
+  if (instance?.isConnected) {
+    instance?.disconnect();
+  }
+};
+
+export const subscribe: Subscribe = ({ instance, disconnect }) => {
+  instance.on('disconnect', async () => disconnect());
```
Comment (RanGojo) [inline wallets/provider-solflare-snap/src/index.ts:42]:
```suggestion
  instance?.on('disconnect', async () => disconnect());
```

**Review by RanGojo — APPROVED** [review]
@RyukTheCoder Please squash your commit to merge  it.

**Pull request merged**